### ISSUE TITLE
fix: systematic neural network backward pass and training bugs across 15+ models

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1049]
## Overview

434 test failures across ~15 neural network models in `ModelFamilyTests.NeuralNetworks` (771/1205 pass = 64%). Most failures trace back to incorrect backward pass gradient computations that don't match the original research papers.

**Pattern discovered:** The GRU fix (PR #1048) found that the reset gate gradient was using h_t instead of h_{t-1} — a single variable reference bug that caused complete training divergence. The same class of bug (wrong tensor in backward pass) likely exists in the other failing models.

### Critical (20/21 or 15/17 failures — model is completely broken)

| Model | Pass/Total | Paper | Root Cause Category |
|-------|-----------|-------|-------------------|
| **ConditionalGAN** | 1/21 | Mirza & Osindero 2014 | Needs custom test setup (latentDim+numClasses input) |
| **DenseNetNetwork** | 2/17 | Huang et al. 2017 | Training/backward failures with 224x224 conv |
| **MobileNetV3Network** | 2/17 | Howard et al. 2019 | Training/backward failures with 224x224 conv |
| **MeshCNN** | 2/17 | Hanocka et al. 2019 | Needs edge adjacency before Forward |

### High (8-11/17 failures — training broken, forward works)

| Model | Pass/Total | Paper | Root Cause Category |
|-------|-----------|-------|-------------------|
| **UnifiedMultimodalNetwork** | 6/17 | — | Training + output shape issues |
| **DeepBoltzmannMachine** | 7/17 | Salakhutdinov & Hinton 2009 | Output shape + training divergence |
| **CapsuleNetwork** | 8/17 | Sabour et al. 2017 | Training divergence (Squash backward?) |
| **VoxelCNN** | 8/17 | Maturana & Scherer 2015 | Training divergence with 3D conv |
| **HTMNetwork** | 8/17 | Hawkins et al. 2016 | Collapsed output (ignores input) + shape mismatch |
| **GraphAttentionNetwork** | 12/20 | Veličković et al. 2018 | Backward pass rank mismatch (#1043) |
| **AudioVisualEventLocalization** | 8/17 | Tian et al. 2018 | Output shape + training failures |
| **BGE** | 9/17 | Xiao et al. 2023 | Training failures (embedding model, 15min test) |
| **NEATTests** | 8/17 | Stanley & Miikkulainen 2002 | Needs genome configuration |

### Low (1-2/17 failures — mostly working)

| Model | Pass/Total | Failing Test | Root Cause |
|-------|-----------|-------------|-----------|
| **GRUNeuralNetwork** | 16/17 | Training_ShouldReduceLoss | **FIXED in PR #1048** — reset gate gradient bug |
| **LiquidStateMachine** | 16/17 | 1 flaky test | May be non-deterministic reservoir |
| **SpikingNeuralNetwork** | 16/17 | 1 test | Surrogate gradient edge case |
| **QuantumNeuralNetwork** | 15/17 | Clone + 1 other | Parameter-shift rule limitation |

### Passing (17/17 — no issues)

OctonionNeuralNetwork, HyperbolicNeuralNetwork, FeedForwardNeuralNetwork, ResidualNeuralNetwork, many others.

## How to Fix Each Model (Pattern from GRU Fix)

The GRU fix (PR #1048) provides the template:

### Step 1: Read the Paper
Find the model's `[ModelPaper]` attribute in the source file. Read the paper's equations for the forward AND backward pass.

### Step 2: Trace the Backward Pass  
In `src/NeuralNetworks/Layers/{ModelLayer}.cs`, find `BackwardManual()` and verify each gradient computation against the paper:
- Are the correct cached tensors being used? (h_{t-1} vs h_t was the GRU bug)
- Are gate derivatives correct? (sigmoid' = σ(1-σ), tanh' = 1-tanh²)
- Is gradient clipping applied for recurrent models?

### Step 3: Check the Test Configuration
In `tests/AiDotNet.Tests/ModelFamilyTests/NeuralNetworks/{Model}Tests.cs`:
- Does `InputShape` match the model's default constructor?
- Does `OutputShape` match the model's actual output?
- Does `CreateNetwork()` provide required configuration?

### Step 4: Run Tests
```bash
dotnet test tests/AiDotNet.Tests/AiDotNetTests.csproj --no-build -f net10.0 \
  --filter "FullyQualifiedName~{Model}Tests" -- xUnit.MaxParallelThreads=1
```

### ConditionalGAN (Mirza & Osindero 2014)
- **File:** `src/NeuralNetworks/ConditionalGAN.cs`
- **Issue:** Default constructor creates GAN expecting `latentDim+numClasses` input (100+10=110), test passes `[1,4]`
- **Fix:** Override `InputShape => [110]` and `OutputShape` in test. May also need generator/discriminator architecture config.

### HTMNetwork (Hawkins et al. 2016)
- **File:** `src/NeuralNetworks/HTMNetwork.cs`
- **Issue:** `DifferentInputs` test fails — output is identical for different inputs. Forward pass ignores input values.
- **Fix:** Check SpatialPoolerLayer/TemporalMemoryLayer forward pass. The spatial pooler may be using column indices that don't depend on input values.

### GraphAttentionNetwork (Veličković et al. 2018)
- **File:** `src/NeuralNetworks/GraphAttentionNetwork.cs`
- **Issue:** Backward produces wrong rank gradient tensor. See #1043.
- **Fix:** Check GraphAttentionLayer backward — ensure output gradient rank matches what LayerBase expects.

### DeepBoltzmannMachine (Salakhutdinov & Hinton 2009)
- **File:** `src/NeuralNetworks/DeepBoltzmannMachine.cs`
- **Issue:** Output shape mismatch + training divergence. DBM uses contrastive divergence, not standard backprop.
- **Fix:** Check output dimension and ensure CD-k training produces correct gradient direction.

### CapsuleNetwork (Sabour et al. 2017)
- **File:** `src/NeuralNetworks/Layers/CapsuleLayer.cs`
- **Issue:** Training diverges. May be Squash activation Jacobian or routing-by-agreement gradient.
- **Fix:** Verify Squash Jacobian against paper equations. Check if routing coefficients are properly backpropped.

### DenseNet/MobileNetV3 (Huang 2017 / Howard 2019)
- **File:** `src/NeuralNetworks/DenseNetNetwork.cs`, `MobileNetV3Network.cs`
- **Issue:** 15/17 failures with 224x224 input. Only ForwardPass and Architecture tests pass.
- **Fix:** Likely backward pass issue in conv/batchnorm/pooling chain. Check gradient shapes through the network.

## Test Commands

```bash
# Run all NN model family tests
dotnet test tests/AiDotNet.Tests/AiDotNetTests.csproj -f net10.0 \
  --filter "FullyQualifiedName~ModelFamilyTests.NeuralNetworks" -- xUnit.MaxParallelThreads=1

# Run specific model
dotnet test tests/AiDotNet.Tests/AiDotNetTests.csproj -f net10.0 \
  --filter "FullyQualifiedName~{Model}Tests" -- xUnit.MaxParallelThreads=1

# Run specific test
dotnet test tests/AiDotNet.Tests/AiDotNetTests.csproj -f net10.0 \
  --filter "FullyQualifiedName~{Model}Tests.Training_ShouldReduceLoss" -- xUnit.MaxParallelThreads=1
```

## Related Issues
- #1039 — GRU fixed (PR #1048), GraphNN still open
- #1043 — GraphAttentionNetwork backward rank mismatch
- #1042 — ConditionalGAN, MeshCNN, NEAT custom test setup
- #1040 — HTMNetwork collapsed output
